src/pybind/mgr/dashboard/controllers/frontend_logging.py

```diff
@@ -1,10 +1,16 @@
+from __future__ import absolute_import
+
+import logging
+
 from . import UiApiController, BaseController, Endpoint
-from .. import logger
+
+
+logger = logging.getLogger('frontend.error')
 
 
 @UiApiController('/logging', secure=False)
-class Logging(BaseController):
+class FrontendLogging(BaseController):
 
     @Endpoint('POST', path='js-error')
     def jsError(self, url, message, stack=None):  # noqa: N802
-        logger.error('frontend error (%s): %s\n %s\n', url, message, stack)
+        logger.error('(%s): %s\n %s\n', url, message, stack)
```

Comment:
If we remove the newlines then it becomes hard to read the stack trace. If I need to grep for frontend errors, I just grep for `frontend.error` (the name of the logger) and then I analyze the stack trace of each occurrence individually.

Even in python, when you call `logger.exception()` it will print the stack trace with newlines in the log file.

Reply:
Multi-line logs are often regarded as a bad practice: they are harder to parse, pollute log formatting, may break some log collectors, and don't necessarily make information clearer (please see my example below).

The stack trace is easier to read, but that doesn't mean it's even useful. The following is a real traceback we grep'ed from our mgr logs:
```
2019-10-31T10:50:55.168+0000 7fd5b85cb700  0 mgr[dashboard] dashboard_exception_handler 
Traceback (most recent call last):
  File "/ceph/src/pybind/mgr/dashboard/services/exception.py", line 88, in dashboard_exception_handler
    return handler(*args, **kwargs)
  File "/usr/lib/python3.7/site-packages/cherrypy/_cpdispatch.py", line 60, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/ceph/src/pybind/mgr/dashboard/controllers/__init__.py", line 657, in inner
    ret = func(*args, **kwargs)
  File "/ceph/src/pybind/mgr/dashboard/controllers/__init__.py", line 850, in wrapper
    return func(*vpath, **params)
  File "/usr/lib64/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
```
It's useless as I had to trim it upfront based on how many lines I expected it'd take, and arbitrarily chose 10 (`grep -i -A10 traceback`), but in this very case that's not enough. And I cannot guess how many lines a traceback is going to take, because that depends on the nesting level of the exception.

On the other hand, if we log the traceback in a single-line, serialized with some special char other than newline (e.g.: '\0' or '|') or with JSON:
```
2019-10-31T10:50:55.168+0000 7fd5b85cb700  0 mgr[dashboard] dashboard_exception_handler: Traceback (most recent call last):|  File "/ceph/src/pybind/mgr/dashboard/services/exception.py", line 110, in handle_rados_error|    yield|...|File "rados.pyx", line 573, in rados.requires.wrapper.validate_func|  File "rados.pyx", line 1257, in rados.Rados.open_ioctx| File "rados.pyx", line 751, in rados.Rados.require_state| rados.RadosStateError: RADOS rados state (You cannot perform that operation on a Rados object in state configuring.)
```
with something like `tail -f mgr.x.log | grep Traceback | tr '|' '\0'`, I'll get:
```
Traceback (most recent call last):                                                                   
  File "/ceph/src/pybind/mgr/dashboard/services/exception.py", line 110, in handle_rados_error       
    yield                                                                                            
(...)                                            
  File "rados.pyx", line 573, in rados.requires.wrapper.validate_func                                
  File "rados.pyx", line 1257, in rados.Rados.open_ioctx                                             
  File "rados.pyx", line 751, in rados.Rados.require_state                                           
rados.RadosStateError: RADOS rados state (You cannot perform that operation on a Rados object in state configuring.)
```

The default `logger.exception` formatter can be easily modified to use single-line tracebacks with [this recipe](https://docs.python.org/2.7/howto/logging-cookbook.html#customized-exception-formatting).

Reply:
I don't disagree with your point but I also don't fully agree. I think this really depends on how the logs are going to be processed (either by a human, or by an automatic collector). 

Maybe we should make this optional, through a setting, where the user can set if it wants one-liner log messages, or multi-line.
Since, the current behavior of ceph logs is multi-line, and I'm not changing that behavior in this PR, we could implement the one-line option in another PR.

Reply:
:+1: for doing this in a separate PR with an option to control it.